File: Project/plane.js
import Vector3D from './vectors.js';
import Matrix3D from './matrix.js';
import Quaternion from './quaternion.js';
import Line3D from './line.js';

class Plane3D{
    
    constructor(a, b, c, d){
        if (arguments.length === 4 && typeof a === 'object' && a instanceof Vector3D) {
            let v1 = new Vector3D();
            v1 = b;
            v1.subtract(a);
            let v2 = new Vector3D();
            v2 = c;
            v2.subtract(b);
            this.normal = v1.crossProduct(v2).normalize();
            let negativeNormal = new Vector3D();
            negativeNormal.scalarMultiply(-1);

            this.a = this.normal.x;
            this.b = this.normal.y;
            this.c = this.normal.z;
            this.d = negativeNormal.dotProduct(a);
            
        } else {
            this.a = a;
            this.b = b;
            this.c = c;
            this.d = d;
            this.normal = new Vector3D(a, b, c);
        }
    }
    
    intersectionWithPlane(plane) {
        const line = new Line3D();
        line.dir = this.normal.crossProduct(plane.normal);

        const dirLen = line.dir.calculateLength() * line.dir.calculateLength();
        if (dirLen === 0.0) {
            return new Line3D(new Vector3D(0, 0, 0), new Vector3D(0, 0, 0));
        }

        const value = (plane.b * this.d - this.b * plane.d) / (this.b * plane.c - plane.b * this.c);
        line.p = new Vector3D(0,0,0);
        line.p.x = 0;
        line.p.y = (-this.c * value - this.d) / this.b;
        line.p.z = value;

        return line;
    }

    angleWithPlane(plane) {
        const v4 = new Vector3D(this.normal.x, this.normal.y, this.normal.z).normalize();
        const v5 = new Vector3D(plane.normal.x, plane.normal.y, plane.normal.z).normalize();

        const dotProduct = v4.dotProduct(v5);
        const magnitudeProduct = v4.calculateLength() * v5.calculateLength();

        if (magnitudeProduct === 0) {
            console.log("One of the planes has zero magnitude.");
            return null;
        }

        const angleRad = Math.acos(dotProduct);
        const angle = 180 - (180 / Math.PI) * angleRad;

        return angle;
    }

    toString() {
        return `${this.a} ${this.b} ${this.c} ${this.d}`;
    }
}

export default Plane3D;

File: Zad6Cpp/src/Plane.cpp
#include "../lib/Plane.h"

Plane::Plane() {}

Plane::Plane(float a, float b, float c, float d)
{
    this->a = a;
    this->b = b;
    this->c = c;
    this->d = d;
    Vector vec{a,b,c};
    normal = vec;
}

Plane::Plane(Vector A, Vector B, Vector C, Vector D) {
    Vector v1 = B;
    v1.sub(A);
    Vector v2 = C;
    v2.sub(B);
    normal = v1.cross(v2);
    normal.mul( - 1);
    Vector negativeNormal = normal;
    negativeNormal.mul(-1);
    a = normal.x;
    b = normal.y;
    c = normal.z;
    d = negativeNormal.dotProduct(A);
}

Line Plane::IntersectionWithPlane(Plane plane) {
    Line line;
    line.direction = normal.cross(plane.normal);

    float dirLen = line.direction.length() * line.direction.length();
    if (dirLen == 0.0f) {
        return Line({0,0,0},{0,0,0});
    }

    float value = (plane.b * this->d - this->b * plane.d) / (this->b * plane.c - plane.b * this->c);
    line.point.x = 0;
    line.point.y = (-this->c * value - this->d) / this->b;
    line.point.z = value;

    return line;
}


float Plane::AnglePlane(Plane plane) {
    Vector v4(normal.x, normal.y, normal.z);
    Vector v5(plane.normal.x, plane.normal.y, plane.normal.z);

    v4.normalize();
    v5.normalize();

    float dotProduct = v4.dotProduct(v5);

    float angleRad = std::acos(dotProduct);
    float angle = 180 - angleRad * 180 / M_PI;

    return angle;
}

std::ostream& operator<<(std::ostream& strm, const Plane& plane) {

    return strm << plane.a << " " <<plane.b << " " << plane.c << " " << plane.d;
}